### PR TITLE
Implement script repeating self nomination

### PR DESCRIPTION
**Pull request by HoOngEe**

### CODE REVIEW

File: auto-self-nominate/.gitignore [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+# Sonarlint plugin
+.idea/sonarlint
+
+# End of https://www.gitignore.io/api/node,webstorm
+.idea
+lib
+
+/build/
+lastBlockNumber.*
```
Comment (sgkim126) [inline auto-self-nominate/.gitignore]: It is not a good habit to copy the configuration you don't know.

File: auto-self-nominate/index.ts [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        case "mainnet":
+            return "cc";
+        default:
+            throw Error("Invalid server configuration");
+    }
+}
+
+const SERVER: string = (() => {
+    const server = process.env.SERVER || "beagle";
+    if (["beagle", "corgi", "mainnet"].indexOf(server) >= 0) {
```
Comment (sgkim126) [inline auto-self-nominate/index.ts]: It's not flexible.
Please read the threads related to https://github.com/CodeChain-io/cron-jobs/issues/61#issuecomment-501132502

File: auto-self-nominate/index.ts [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+function decodePlatformAddressFromPubkey(buffer: Buffer): PlatformAddress {
+    const pubkey = buffer.toString("hex");
+    return PlatformAddress.fromPublic(pubkey, {
+        networkId: sdk.networkId,
+    });
+}
+
+function decodeNumber(buffer: Buffer): number {
+    return parseInt("0x" + buffer.toString("hex"), 16);
```
Comment (sgkim126) [inline auto-self-nominate/index.ts]: You don't need to add "0x"

File: auto-self-nominate/index.ts [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+async function getCurrentTermId(bestBlockNumber: number): Promise<number | null> {
```
Comment (sgkim126) [inline auto-self-nominate/index.ts]: It's not a comment about this commit, but please make a commit that adds this to SDK.

File: auto-self-nominate/index.ts [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        networkId(SERVER),
+    );
+}
+
+async function needsNomination(
+    address: string,
+    nominationFrequencyLevel: number,
+): Promise<boolean> {
+    const bestBlockNumber = await sdk.rpc.chain.getBestBlockNumber()!;
+    const candidates: Map<string, number> = await (async () => {
```
Comment (sgkim126) [inline auto-self-nominate/index.ts]: Usually, you don't need to use `Map`, if the type of key is `string` and you don't iterate over it.

File: auto-self-nominate/index.ts [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                decodePlatformAddressFromPubkey(encodedPubkey).value,
+                decodeNumber(encodedNominationEnd),
+            ),
+        );
+        return retval;
+    })();
+    const nominationEndAt = candidates.get(address);
+    if (nominationEndAt == null) {
+        throw new Error(`SelfNominate first with specific deposit vakue before repeating`);
+    } else {
```
Comment (sgkim126) [inline auto-self-nominate/index.ts]: You don't need to use `else`. The if block always throws an error.

File: auto-self-nominate/index.ts [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+        const currentTermId = (await getCurrentTermId(bestBlockNumber))!;
+        return nominationEndAt - currentTermId < nominationFrequencyLevel;
+    }
+}
+
+async function main() {
+    const accountAddress = getConfig("ACCOUNT_ADDRESS");
+    const passphrase = getConfig("PASSPHRASE");
+    const metadata = getConfig("METADATA");
+    const repeatedAdditionalDeposit = new U64(parseInt(getConfig("ADDITIONAL_DEPOSIT"), 10) || 0);
```
Comment (sgkim126) [inline auto-self-nominate/index.ts]: I think an addtional deposit is not convenient to use.
Could you change it to the target deposit?

File: auto-self-nominate/index.ts (hunk 3 above)
Comment (HoOngEe) [reply auto-self-nominate/index.ts]: Fixed

File: auto-self-nominate/index.ts (hunk 5 above)
Comment (HoOngEe) [reply auto-self-nominate/index.ts]: I fixed the type of candidates to `{ [key: string] : number}`

File: auto-self-nominate/index.ts (hunk 6 above)
Comment (HoOngEe) [reply auto-self-nominate/index.ts]: Fixed

File: auto-self-nominate/index.ts (hunk 7 above)
Comment (HoOngEe) [reply auto-self-nominate/index.ts]: Fixed

File: auto-self-nominate/index.ts (hunk 2 above)
Comment (HoOngEe) [reply auto-self-nominate/index.ts]: Fixed to get `NETWORK_ID` and `RPC_URL` from environment variables.

File: auto-self-nominate/index.ts [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+        ([encodedPubkey, encodedDeposit, encodedNominationEnd]) =>
+            (retval[decodePlatformAddressFromPubkey(encodedPubkey).value] = [
+                decodeNumber(encodedNominationEnd),
+                decodeNumber(encodedDeposit),
+            ]),
+    );
+    return retval;
+}
+
+async function supplementaryNoimationDeposit(
```
Comment (sgkim126) [inline auto-self-nominate/index.ts]: This function does two things. One is checking whether it should send a nominate transaction, and the other is calculating the deposit. How about renaming the function or split the function?

File: auto-self-nominate/.gitignore (hunk 1 above)
Comment (HoOngEe) [reply auto-self-nominate/.gitignore]: I trimmed `.gitignore`

File: auto-self-nominate/index.ts (hunk 8 above)
Comment (HoOngEe) [reply auto-self-nominate/index.ts]: I split the function into two different functions.

File: auto-self-nominate/index.ts [hunk 9]
```diff
@@ -0,0 +1,169 @@
+import { U64 } from "codechain-primitives/lib";
+import { SDK } from "codechain-sdk";
+import { PlatformAddress } from "codechain-sdk/lib/core/classes";
+import { Custom } from "codechain-sdk/lib/core/transaction/Custom";
+
+const RLP = require("rlp");
+
+const ACTION_TAG_SELF_NOMINATE = 4;
+const STAKE_ACTION_HANDLER_ID = 2;
+const NEED_NOMINATION_UNDER_TERM_LEFT = 5;
```
Comment (sgkim126) [inline auto-self-nominate/index.ts]: How about making it configurable with the default value?

Comment (sgkim126) [reply auto-self-nominate/index.ts]: And I think 2 is enough for the default value.

File: auto-self-nominate/index.ts [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+async function main() {
+    const accountAddress = getConfig("ACCOUNT_ADDRESS");
+    const passphrase = getConfig("PASSPHRASE");
+    const metadata = getConfig("METADATA");
+    const targetDeposit = parseInt(getConfig("TARGET_DEPOSIT"), 10);
+
+    setInterval(async () => {
+        const bestBlockNumber = await sdk.rpc.chain.getBestBlockNumber()!;
+        const info = await getCandidatesInfo(bestBlockNumber);
+        if (needsNomination(info, bestBlockNumber, accountAddress)) {
```
Comment (sgkim126) [inline auto-self-nominate/index.ts]:
```js
if (await needsNomi...
```

Comment (sgkim126) [reply auto-self-nominate/index.ts]: How about setting https://palantir.github.io/tslint/rules/strict-boolean-expressions/?

Comment (HoOngEe) [reply auto-self-nominate/index.ts]: Fixed and set the rule `strict-boolean-expression` true.

File: auto-self-nominate/index.ts (hunk 9 above)
Comment (HoOngEe) [reply auto-self-nominate/index.ts]: made it configurable with the default value 2

**Pull request merged**